feat: add common API module and util function

RookieAND opened the pull request:
## Jira Issue 번호

[#57](https://gongback.atlassian.net/browse/WEEDA-57)

## 작업 내용

- 서버와의 통신을 위한 axios 라이브러리 추가
- 클라이언트에서 공용으로 사용할 API 모듈에서 사용될 Type 지정
   - APIResponse
   - APIError
- 클라이언트에서 공용으로 사용할 API 유틸 함수 제작
   - getAsync
   - postAsync
   - patchAsync
   - deleteAsync
- 공통된 config 설정을 위한 axios 인스턴스 추가 및 적용
	- baseURL, responseType 설정
- 서버와의 통신에 필요한 baseURL 를 상수로 만들어 보관

## Checklist

- [X] Code Review 요청
- [X] PR 제목 commit convention에 맞는지 확인
- [X] Label 설정
- [X] Jira 코드 리뷰 상태로 변경

ballsona committed:
WEEDA-30 [feat]: add react-native-svg package and svg files

ballsona committed:
WEEDA-30 [feat]: add functions for responsive style

ballsona committed:
WEEDA-30 [feat]: add AuthForm component

ballsona committed:
WEEDA-30 [feat]: add AuthHeader component

ballsona committed:
WEEDA-30 [feat]: add LoginTemplate component and LoginScreen

ballsona committed:
WEEDA-30 [refactor]: change atom directory to common

ballsona committed:
WEEDA-30 [feat]: add types about Auth

ballsona committed:
WEEDA-30 [fix]: fix storybook decorator for layout

RookieAND committed:
WEEDA-57 [feat]: add axios package

API 통신에 필요한 axios 라이브러리를 설치하였습니다.

RookieAND committed:
WEEDA-57 [fix]: modify axios version by project convention

패키지의 특정 버전을 명시하기 위해 ^ 기호를 제거하여 라이브러리를 재적용하였습니다.

RookieAND committed:
WEEDA-57 [feat]: add some type of API module

공용 API 함수에 사용될 type을 우선적으로 지정하여 등록하였습니다.

RookieAND committed:
WEEDA-57 [feat]: add API util function and axios instance

API 호출에 범용적으로 사용될 유틸 함수와 axios 인스턴스를 생성했습니다.

RookieAND committed:
WEEDA-57: [feat]: add backend server url to const variable

통신을 진행할 백엔드 URL을 상수로 저장하여 API 호출에 사용하도록 하였습니다.

RookieAND commented:
@ballsona 아직 WEEDA-30 브랜치를 dev 에 merge 하지 않으셔서 해당 브랜치를 기반으로 코드를 작성하였습니다.
먼저 WEEDA-30 브랜치를 merge 하신 후에 코드 리뷰를 진행해주시면 즉시 이 브랜치도 merge 하겠습니다.

RookieAND committed:
WEEDA-57 [feat]: add type of APITypes in index.ts

API 관련 타입을 index.ts 에서 export 하여 경로를 축약하였습니다.

RookieAND committed:
WEEDA-57 [fix]: add endpoint in API base url

API_URL 변수에 공통으로 들어가는 엔드포인트를 추가하였습니다.

RookieAND commented:
연이어서 커밋을 진행하니 각 커밋에 대한 build 도 개별적으로 진행하네요.. 앞으로는 PR을 날릴 때 놓친 부분이 없는지 다시 한번 체크해보도록 하겠습니다.

ballsona commented on src/constants/types/APITypes.ts:
```diff
@@ -0,0 +1,19 @@
+/**
+ * API 호출 결과 (성공, 실패) 를 나타내는 type
+ * @param T 요청 성공 시 인계 받은 데이터 타입
+ */
+export type APIResponse<T> = Promise<
+  { isSuccess: true; result: T } | { isSuccess: false; result: APIError }
+>;
+
+/**
+ * API 호출 과정에서 발생한 에러와 관련된 interface
+ * @param code 백엔드 측에서 전송한 HTTP Status Code
+ * @param msg 백엔드 측에서 전송한 에러 관련 메세지
+ * @param data 백엔드 측에서 전송한 에러 관련 데이터
+ */
+export interface APIError {
+  code: number;
+  msg: string;
+  data?: any;
+}
```
`ApiResponse`, `ApiError` 으로 CamelCase를 사용하는게 통일성 있을 것 같은데 어떻게 생각하시나요 ??

ballsona replied:
앗 constants 내 파일명들은 소문자로 시작하는걸로 통일해서 `apiTypes.ts` ,`authTypes.ts` 로 지정하는게 좋을것 같아요.. 제가 잘못 설정한 것 같네요 죄송함다 🥹

RookieAND replied:
camelCase 를 준수하여 코드를 작성하는 게 맞는 것 같습니다. 습관적으로 API를 대문자로 쓰다보니 벌어진 문제 같습니다. 바로 수정하도록 하겠습니다.

ballsona commented on src/constants/types/APITypes.ts:
```diff
@@ -0,0 +1,19 @@
+>;
+
+/**
+ * API 호출 과정에서 발생한 에러와 관련된 interface
+ * @param code 백엔드 측에서 전송한 HTTP Status Code
+ * @param msg 백엔드 측에서 전송한 에러 관련 메세지
+ * @param data 백엔드 측에서 전송한 에러 관련 데이터
+ */
+export interface APIError {
+  code: number;
+  msg: string;
+  data?: any;
+}
```
요부분은 함수의 매개변수를 나타내는 `@param` 보다는 요렇게 주석을 다는게 좋을 것 같아요!

```suggestion
export interface APIError {
  /** 백엔드 측에서 전송한 HTTP Status Code */
  code: number;
  /** 백엔드 측에서 전송한 에러 관련 메세지*/
  msg: string;
  /** 백엔드 측에서 전송한 에러 관련 데이터 */
  data?: any;
}
```

RookieAND replied:
오... 그렇군요, 확실히 `@param` 의 경우에는 함수의 매개변수를 의미하는 터라 interface는 어떻게 작성해야 할지 잘 몰랐습니다.
앞으로 다른 프로젝트에서도 인터페이스 관련 주석을 달 때는 다음과 같이 작성을 해야겠네요. 알려주셔서 감사합니다.

ballsona commented on src/apis/API.ts:
```diff
@@ -0,0 +1,141 @@
+ * @param config API 요청과 관련된 config (AxiosRequestConfig)
+ * @returns API 요청 성공과 실패에 따른 객체 (APIResponse)
+ */
+export async function patchAsync<T, D>(
+  url: string,
+  data: D,
+  config?: AxiosRequestConfig,
+): APIResponse<T> {
+  try {
+    const response = await API.patch<T, AxiosResponse<T, D>, D>(url, data, {
+      ...config,
+    });
+    return { isSuccess: true, result: response.data };
+  } catch (err) {
+    return { isSuccess: false, result: handleAPIError(err) };
+  }
+}
```
오 patch 메서드는 개념만 알고 있었는데 덕분에 처음 써볼 것 같아요! :) 👍

ballsona commented on src/apis/API.ts:
```diff
@@ -0,0 +1,141 @@
+ * @param err API 통신 과정에서 발생한 에러 데이터
+ * @returns 클라이언트에게 인계할 에러 객체 (APIError)
+ */
+function handleAPIError(err: unknown): APIError {
```
혹시 `handleApiError` 로 CamelCase 로 함수 및 변수명 통일하는거 어떠실까요..?

RookieAND replied:
이 부분도 API => Api 로 CamelCase를 준수하여 작성하도록 하겠습니다.

ballsona commented on src/apis/API.ts:
```diff
@@ -0,0 +1,141 @@
+  }
+}
+
+/**
+ * PATCH 요청을 처리하는 유틸 API 함수 postAsync
```
요기 `patchAsync` 로 변경 부탁드립니다!

ballsona commented on src/apis/API.ts:
```diff
@@ -0,0 +1,141 @@
+}
+
+/**
+ * PATCH 요청을 처리하는 유틸 API 함수 postAsync
```
요기도 함수명 변경해야할 것 같아용!

RookieAND replied:
복붙의 위험성을 다시금 느끼네요. 말씀하신 주석의 함수명도 일괄적으로 변경하도록 하겠습니다.

ballsona reviewed (changes requested):
광인 작업하시느라 수고하셨습니다! api util 부분은 많이 까다로우셨을텐데 많이 배워갑니다 👍 

다만 feature 브랜치는 꼭 develop 브랜치에서 분기가 되어야 해요. (그래야 광인님 PR에 광인님이 작업하신 코드만 담길 수가 있어요!) 웬만하면 각 feature 마다 독립적인 작업을 담당하기 때문에 develop에서 병합되어도 충돌나지 않을거에요. 다음 작업부터는 요렇게 부탁드릴게요 :)

ballsona commented:
> 연이어서 커밋을 진행하니 각 커밋에 대한 build 도 개별적으로 진행하네요.. 앞으로는 PR을 날릴 때 놓친 부분이 없는지 다시 한번 체크해보도록 하겠습니다.

커밋 이어서 하셔도 괜찮습니다! 저도 수정하느라 많이 한걸요.. ㅎㅎ

RookieAND commented:
코드 리뷰해주셔서 감사합니다. 추후에는 무조건 `develop` 브랜치에서 생성하여 작업을 진행하도록 하겠습니다.
리뷰 중 포인트를 짚어주신 부분은 지금 바로 일괄적으로 수정해보도록 하겠습니다. 감사합니다.

RookieAND committed:
WEEDA-57 [fix]: change some identifiers to follow camelCase

코드에 작성된 식별자들 중 일부를 camelCase에 맞춰 재작성하였습니다.
주석에서 잘못 작성되었던 유틸리티 함수 명도 정확하게 재작성하였습니다.

RookieAND committed:
WEEDA-57 [fix]: change interface's name to follow camelCase

API에 사용되는 인터페이스의 이름을 camelCase를 준수하도록 수정하였습니다.
interface의 JSDoc 또한 함수의 매개변수를 의미하는 @param 을 사용하지 않고 속성에 주석을 두었습니다.

RookieAND committed:
WEEDA-57 [fix]: change types file name to follow camelCase

index.ts 에 적용되었던 기존의 파일 명을 camelCase를 준수한 현재의 이름을 수정하였습니다.

ballsona committed:
Merge branch 'develop' into feature/WEEDA-57-common-api-module
